### PR TITLE
BAPL-1217 Stunner - View all validation errors

### PR DESCRIPTION
**Pull request by tiagodolphine**
[BAPL-1217](https://issues.jboss.org/browse/BAPL-1217) Improve the error messages and show all validation errors that a user needs to know about in order to fix the process model.
[JBPM-7436)](https://issues.jboss.org/browse/JBPM-7436)

- Violations messages split by node ID, before it was an unique message
- Each node with any violation is highlighted in Red
- Messages, in general, were simplified to be easier to read
- Every violation is now associated with the node Id
- Model validation is now working (it has some bug before they were not triggered)
- The validation message popup  was simplified, now Warning/Error just say to look at the panel for details, this avoid to flood the screen with many long messages

TODO: implement the missing tests.

@romartin @LuboTerifaj 
I opened the PR it is missing more tests and some code improvements, but if you can help to review and give suggestions, it is a long PR and I'm afraid we are getting out of time.

Thanks guys !

### CODE REVIEW

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
hey @tiagodolphine , thanks for this job! :+1: 

wondering when I'll be able to give it a try.... :) builds are failing: 
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (validate) on project kie-wb-common-stunner-core-common: You have 2 Checkstyle violations`

**Comment by tiagodolphine** [comment]
Dammit, I got this after rebase I think, anyway I fixed the issues, now it should work @romartin

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
@romartin @wmedvede did you have a chance to take a look on this?
thanks

**Comment by romartin** [comment]
sorry @tiagodolphine , still found no time... but just start looking at it :+1: 
CC @wmedvede @LuboTerifaj

**Comment by romartin** [comment]
@tiagodolphine can the word "WIP" in the PR's title be removed at this point? :) Looks nobody will review...

**Comment by tiagodolphine** [comment]
@romartin removed, but at this point, we need more hands to help on the review, even it is not the final version but at least give suggestions and help to find bugs to have at least a version with some good improvements on this, my opinion is that we can fix all bugs if any, and open new JIRAs as improvements for the things that are not bugs, WDYT ? this may include the unit tests because there will be no time IMO.
thanks

**Comment by romartin** [comment]
I know those above are not bugs, but if it's easy to fix will be a quick win by improving the experience.... :1st_place_medal:

**Comment by tiagodolphine** [comment]
Hi @romartin  thanks for the review!
1. I'll try to get the `name` but this may lead at some conflicts when nodes have the same name, anyway for the user perspective the uuid doesn't make sense, better the name I'll try to get that !
2. yeah I know, but the problem is that we are using the bootstrap Notify component, that has a king of delay/fade out even calling the hide() and setting no effect, this was the most fast I got. If you have any clue how to solve this., I can apply.

Thanks.

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java [hunk 1]
```diff
@@ -192,35 +194,35 @@ public void setContentScrollType(final ScrollType type) {
 
     @Override
     public SessionPresenterView showError(final String message) {
-
+        Notify.hideAll();
         getSettings().setType(kieNotificationCssClass(NotifyType.DANGER));
-        showNotification(translate(SessionPresenterView_Error), message, IconType.EXCLAMATION_CIRCLE);
+        showNotification(translate(SessionPresenterView_Error),
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java]: This causes the `message` argument to be unsued.... IMO is confusing, as the Alerts panel could be not present (standalone/submarine)...

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java [hunk 2]
```diff
@@ -192,35 +194,35 @@ public void setContentScrollType(final ScrollType type) {
...
     }
 
     @Override
     public SessionPresenter.View showWarning(final String message) {
-
+        Notify.hideAll();
         getSettings().setType(kieNotificationCssClass(NotifyType.WARNING));
-        showNotification(translate(SessionPresenterView_Warning), message, IconType.EXCLAMATION_TRIANGLE);
-
+        showNotification(translate(SessionPresenterView_Warning),
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java]: same here, unused `message` argument

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java (hunk 1 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java]: This part is easy, just remove the `message` parameter.

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java (hunk 2 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/presenters/session/impl/SessionPresenterView.java]: This part is easy, just remove the `message` parameter.

**Comment by tiagodolphine** [comment]
@romartin about 
1. yeah I agree, let's try to do it
2. I already tried to disable the animations but it doesn't solve the issue of showing multiples alerts, even hiding them before. I don't think to keep the old implementation even for the standalone makes sense, now that we have multiple messages on each validation, in general, it may trigger lots of alerts popups. One other option is to use other component other than the `Bootstrap Notify`, do you know any other? And one last option is to remove the alterts popup at all for ERROR/WARNING.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java [hunk 3]
```diff
@@ -74,6 +79,10 @@ private void checkViolations(final H canvasHandler,
         final String title = diagram.getMetadata().getTitle();
         final boolean[] valid = {true};
         elementViolations
+                .stream()
+                .filter(v -> v.getDomainViolations().stream().map(DomainViolation::getMessage).anyMatch(StringUtils::nonEmpty)
```
Comment (wmedvede) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java]: A refactoring of this filtering conditions   v.getDomainViolations().stream().map(DomainViolation::getMessage).anyMatch(StringUtils::nonEmpty) would be nice

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java [hunk 4]
```diff
@@ -51,6 +54,11 @@
         this.modelViolations = modelViolations;
         this.domainViolations = domainViolations;
         this.type = type;
+        this.translationService = Optional.empty();
+    }
+
+    public void setTranslationService(StunnerTranslationService translationService) {
```
Comment (wmedvede) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java]: I understand translations are needed, but that every ElementViolation has a reference to the translation service looks a bit strange

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java [hunk 5]
```diff
@@ -112,6 +129,11 @@ public Builder setType(Violation.Type type) {
             return this;
         }
 
+        public Builder setTranslationService(StunnerTranslationService translationService) {
```
Comment (wmedvede) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java]: same here, to buid an and ElementViolation I need to provide a translation service. I see it's an optional, but probably the responsible of making the translation happen should be the component that wants show/paint the ElementViolation somewhere

File: kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/widgets/Notifications.java [hunk 6]
```diff
@@ -184,9 +184,9 @@ public String getValue(final Notification object) {
 
     @SuppressWarnings("unchecked")
     private String getNotificationSourceMessage(final Notification notification) {
-        return notification.getSource().isPresent() ?
-                notification.getSource().get().toString() :
-                "-- No source --";
+        return (String) notification.getSource()
+                .map(s -> notification.getMessage())
+                .orElse("-- No source --");
```
Comment (wmedvede) [inline kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/widgets/Notifications.java]: looks like the Notifications widget has some lack of i18n.

**Comment by romartin** [comment]
Hey @tiagodolphine ,

Well in fact IIRC we already talked about removing the popups that appear for the validation messages, but it should only happens on the workbench environment (project modules), as for the standalone (submarine) modules, we don't have the alerts panel, so we shoul still use the popups or built some other component about it.

Anyway if that's not trivial, due to the work load we've , I'm agree on going ahead as it is, and is necessary, improving this for next release. So don't worry about it for now, let's focus on complete the expected tasks for the release... :+1: 

Thanks!

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java (hunk 3 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/validation/canvas/CanvasDiagramValidator.java]: Yeah, I agree

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java (hunk 4 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java]: Yeah, I agree, I'll remove it, and let to the caller handle the message.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java (hunk 5 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/validation/impl/ElementViolationImpl.java]: Yeah, I agree, I'll remove it, and let to the caller handle the message.

File: kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/widgets/Notifications.java (hunk 6 above)
Comment (tiagodolphine) [reply kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/widgets/Notifications.java]: yeah, actually this class if from the standalone showcase and it doesn't handle i18n at all. Probably it is better to open a JIRA to add i18n on all this class.

**Comment by romartin** [comment]
missing your review plz @LuboTerifaj

**Comment by tiagodolphine** [comment]
I'll update the PR with improvements suggested on the comments, so wait for the new commit plz, I'll do this tomorrow. Anyway, with this it is just missing the unit tests, that will be provided ASAP, I'm giving priority to this now since the roundtrip was moved to 7.5.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
I pushed some changes, from the comments it is missing the `Notify` for the message popup. I'll take a look at this tomorrow, and start the unit tests.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
Pushed a new commit, now there is a single warn notification popup.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
what's the status here guys? can this be merged @LuboTerifaj ? thx!

**Comment by romartin** [comment]
@LuboTerifaj pleeaseeeee! :)

**Comment by LuboTerifaj** [comment]
Can you @tiagodolphine rebase your branch with master please?

**Comment by tiagodolphine** [comment]
@LuboTerifaj thanks for the review, I just rebased the branch.

@romartin once this is green again we are ready to merge it.

**Comment by tiagodolphine** [comment]
Jenkins execute full downstream build

**Comment by tiagodolphine** [comment]
@romartin ok to be merged

**Pull request merged**